File: leetcode/Python/0094-m-binary-tree-inorder-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

"""
Recursion
"""
class Solution_1:
    def inorderTraversal(self, root: TreeNode) -> List[int]:
        res = []
        self.inorder(root, res)
        return res

    def inorder(self, node, res):
        if not node:
            return
        self.inorder(node.left, res)
        res.append(node.val)
        self.inorder(node.right, res)

"""
Non-Recursion
"""
class Solution_2:
    def inorderTraversal(self, root: TreeNode) -> List[int]:
        if not root:
            return []

        dummy = TreeNode(0)
        dummy.right = root
        stack = [dummy]

        res = []
        while stack:
            node = stack.pop()
            if node.right:
                node = node.right
                while node:
                    stack.append(node)
                    node = node.left
            if stack:
                res.append(stack[-1].val)
        return res

File: leetcode/Java/0094-m-binary-tree-inorder-traversal.java
import java.util.Stack;

import javax.swing.tree.TreeNode;

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */

/*
 * Recursion
 */
class Solution_1 {
    public List<Integer> inorderTraversal(TreeNode root) {
        List<Integer> res = new ArrayList<Integer>();
        inorder(root, res);
        return res;
    }

    public void inorder(TreeNode node, List<Integer> res) {
        if (node == null)
            return;
        inorder(node.left, res);
        res.add(node.val);
        inorder(node.right, res);
    }
}

/*
 * Non-Recursion
 */
class Solution_2 {
    public List<Integer> inorderTraversal(TreeNode root) {
        List<Integer> res = new ArrayList<Integer>();
        Stack<TreeNode> stack = new Stack<TreeNode>();

        if (root == null)
            return res;

        TreeNode dummy = new TreeNode(0);
        dummy.right = root;
        stack.push(dummy);

        while (!stack.empty()) {
            TreeNode node = stack.pop();
            if (node.right != null) {
                node = node.right;
                while (node != null) {
                    stack.push(node);
                    node = node.left;
                }
            }
            if (!stack.empty())
                res.add(stack.peek().val);
        }

        return res;
    }
}